### PR TITLE
KEYCLOAK-7675 Support for Device Authorization Grant

### PR DESCRIPTION
**Pull request by splatch**
This is @wadahiro 's work rebased on top of current master.

I've tested this work with 6.0 release and it did work. I am able to commit couple of days for this if it still needs work.

### CODE REVIEW

**Comment by stianst** [comment]
Haven't had a chance to look at this, but if memory serves me correctly one of the things we hadn't agreed on in the design discussion was how to deal with brute-force guessing of the device code. How is that handled here?

**Comment by stianst** [comment]
@wadahiro what was outstanding in your branch, or the reason why you didn't send a PR?

**Comment by wadahiro** [comment]
@stianst 

I'm sorry that I didn't have enough spare time to continue the development.

> we hadn't agreed on in the design discussion was how to deal with brute-force guessing of the device code.

I think we agreed on the following discussion that we don't implement rate limiting to protect the brute-force attacks:

https://github.com/keycloak/keycloak-community/pull/6#discussion_r294637793


> what was outstanding in your branch, or the reason why you didn't send a PR?

I think my original branch needs to improve on the following points:

1. Make consent optional

In the current implementation, the consent screen is always shown. It is implemented by the following code:

https://github.com/keycloak/keycloak/pull/6992/files#diff-e803ffc6e60d9835541e184a9da45e0aR935

As you can see in the discussion thread below in the design document, I think this should be optional.

https://github.com/keycloak/keycloak-community/pull/6#discussion_r282030710


2. Implement the first page when verification_uri_complete is used

As you can see in the discussion thread below, the first page when verification_uri_complete is used is not implemented in my implementation yet.

https://github.com/keycloak/keycloak-community/pull/6#discussion_r284106463

**Comment by splatch** [comment]
>     1. Make consent optional
> In the current implementation, the consent screen is always shown. It is implemented by the following code:
> https://github.com/keycloak/keycloak/pull/6992/files#diff-e803ffc6e60d9835541e184a9da45e0aR935
> As you can see in the discussion thread below in the design document, I think this should be optional.
> [keycloak/keycloak-community#6 (comment)](https://github.com/keycloak/keycloak-community/pull/6#discussion_r282030710)

Is there any particular reason why we should not use "display consent" option which is set at client level? Shall we make another option with "display consent for devices" if same client id is used for different flows (which I doubt)? Or it shows up with different devices, when second one is "consented"? I saw there was a "Display consent always" which seems to be just a variation of condition @wadahiro already pointed.

>     2. Implement the first page when verification_uri_complete is used
> As you can see in the discussion thread below, the first page when verification_uri_complete is used is not implemented in my implementation yet.
> [keycloak/keycloak-community#6 (comment)](https://github.com/keycloak/keycloak-community/pull/6#discussion_r284106463)

All right, so the missing point is accepting the code which is sent in URL. That is something I can bring to the table. :-)

Above topics do not seem to be huge challenges thus, if there is clear indication what should be done for first point I should be able to make it happen, or leave it to Hiroyuki, if he wish to complete PR.

**Review by pedroigor — CHANGES_REQUESTED** [review]
@splatch @wadahiro That is great !

Spec-wise I think as initial implementation, it covers most of the specs.

Implementation-wise, I've only a few considerations, mainly around the storage for device code. Do we really need that?

As an alternative, we could create an authentication session when the device sends an authorization request. The root session associated with this authentication session will have some ID that allows us to cross-reference the code returned to the device and the user code. For instance, the user code used as the id for the root authentication session.

IMO, leverage authentication sessions would be ideal instead of having all that to just support device grant.

**Comment by pedroigor** [comment]
@splatch @wadahiro if you guys are OK with my comments I have something working based on what I have suggested. If you don't have time to work on this, I can take this one and submit another PR for your review.

**Comment by splatch** [comment]
@pedroigor I am booked this week so earliest time I can start working on improvements in this area is next week.

With regard to your proposal device code storage is a tricky one. I think that iterating over sessions simplify code from component point of view, however makes it harder to understand. Given that there is no standard "session processing" interface we end up with plenty of auth notes distributed around number of places. I know it works in other places. The major benefit I see in your proposal is replication of sessions in clustered environments which we get for free so documentation for clustered and multi-dc deployments remains untouched.

Best,
Łukasz

**Comment by pedroigor** [comment]
@splatch Exactly. The fact that we would be leveraging what we already have authentication sessions.

In regards to the usage of sessions and notes, we can implement a very simple API/utility that abstracts and centralizes how we would use those sessions.

Basically, we would be working with a single authentication session.

**Comment by splatch** [comment]
@pedroigor I was thinking of this lately and come to point where I realized that code validation should be rather part of dedicated authorization flow. Currently we have a flow for direct grant but we dont have that luxury for device flow.
Since device might send additional params to device code endpoint and it can also return customized link to device. Flexibility of flows would be desired in order to make this customizable.

File: services/src/main/java/org/keycloak/protocol/oidc/endpoints/OAuth2DeviceAuthorizationEndpoint.java [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+        if (request.getLoginHint() != null) authenticationSession.setClientNote(OIDCLoginProtocol.LOGIN_HINT_PARAM, request.getLoginHint());
+        if (request.getPrompt() != null) authenticationSession.setClientNote(OIDCLoginProtocol.PROMPT_PARAM, request.getPrompt());
+        if (request.getIdpHint() != null) authenticationSession.setClientNote(AdapterConstants.KC_IDP_HINT, request.getIdpHint());
+        if (request.getClaims()!= null) authenticationSession.setClientNote(OIDCLoginProtocol.CLAIMS_PARAM, request.getClaims());
+        if (request.getAcr() != null) authenticationSession.setClientNote(OIDCLoginProtocol.ACR_PARAM, request.getAcr());
+        if (request.getDisplay() != null) authenticationSession.setAuthNote(OAuth2Constants.DISPLAY, request.getDisplay());
+
+        if (request.getAdditionalReqParams() != null) {
+            for (String paramName : request.getAdditionalReqParams().keySet()) {
+                authenticationSession.setClientNote(LOGIN_SESSION_NOTE_ADDITIONAL_REQ_PARAMS_PREFIX + paramName, request.getAdditionalReqParams().get(paramName));
```
Comment (splatch) [inline services/src/main/java/org/keycloak/protocol/oidc/endpoints/OAuth2DeviceAuthorizationEndpoint.java:235]: These client notes are useful, however as far I was able to test, it is not possible to make any use of them. The initial session created by device is valid only during the generation of the code. When user opens verification URI he creates completely new session. The URI he launches has no additional request parameters. In current implementation there is no way to correlate session created by device with one which is continued by user who confirms the code.

Comment (pedroigor) [reply services/src/main/java/org/keycloak/protocol/oidc/endpoints/OAuth2DeviceAuthorizationEndpoint.java:235]: Spec-wise is it expected the client to send additional parameters?

If that is the case, I think we could just have those available through the root session.

**Comment by pedroigor** [comment]
@splatch I'm not sure about having an authentication flow for device. The reason is that we basically authenticate based on the code. Maybe moving the actual code validation to a specific Client Authenticator makes more sense. Like we have for secret, jwt, etc.

IMO, this PR is pretty much OK except for the points I raised in regards to authentication session.

@stianst @mposolda I did some tests using authentication sessions and it seems to work fine for device grant so that we could avoid having a specific cache layer only for device. Wdyt?

**Comment by splatch** [comment]
> @splatch I'm not sure about having an authentication flow for device. The reason is that we basically authenticate based on the code. Maybe moving the actual code validation to a specific Client Authenticator makes more sense. Like we have for secret, jwt, etc.

I believe it makes a lot of sense to reuse this logic and SPI, cause in production deployments first step must be made by pre-authenticated device with a valid certificate. Current implementation leaves no option for that.
WRT session code I been working on that part too, to simply pass an imaginary device serial number from `/device` to user authorization flow, so I could validate additionally if user is permitted to authorize specific device. It works now for me, but I will be able to test this again once your session code is in place.

**Comment by pedroigor** [comment]
@splatch For client cert auth you don't actually need a flow, but just configure your client to use X509 certs for authenticating.

But I see your point. However, I would keep for now the basics and improve later on.

Regarding the session code, here it is what I did when first reviewing these changes. See https://github.com/pedroigor/keycloak/tree/KEYCLOAK-7675.

Some bits like the pooling, etc, are missing because I did port from the code I removed (the new cache layer).

**Comment by DmitryZagr** [comment]
Please tell me when it is planned to introduce this functionality?

**Comment by splatch** [comment]
I am still testing the earlier build in our project. It is pre-production environment with small test load. I hope to get another environment set up before end of June where I will be able to verify recent changes made by Pedro.

**Comment by DmitryZagr** [comment]
any updates?

**Comment by rmoqpd5t** [comment]
I tested this PR using endpoint /auth/realms/<realm>/protocol/openid-connect/device/auth with successful result. Bravo. If you need further testing, please let me know.

**Comment by DmitryZagr** [comment]
What is the fundamental problem of this PR? For example in https://github.com/wso2/product-is this feature was implemented a year ago. This feature is very important for many devices with input restrictions :(

**Comment by pedroigor** [comment]
This PR is waiting for https://github.com/keycloak/keycloak/pull/7679. We should be re-using some key code from there. We definitely want device grant.

**Comment by splatch** [comment]
> This PR is waiting for #7679. We should be re-using some key code from there. We definitely want device grant.

I see that other PR is still going over reviews. Looking at it shape  this one has to be written from scratch to fit into `DecoupledAuthenticationRequest`. I get that you want to unify code, but this thing was ready to be merged, passed over community reviews and design. The ciba stuff could be integrated with necessary refactor. Instead we have almost entire year waiting for final push.

As a side note - token endpoint again grows by another 200 lines of code. It slowly turns into an pile of dissociated grants which will soon require another refactor. :-/

**Comment by pedroigor** [comment]
@splatch Yep, you are right. Discussions are happening to create an SPI for grant types. As well as how we should deal with extensions to protocols. As you noticed, not only token endpoint but usually you need to change a lot of other places.

We won't need to rewrite this one - we did a lot of work here. The type you mentioned is a CIBA-specific thing. The work here is not in vain, but it would benefit from some key aspects from there.

Let us re-open this one as it is still an important piece of work (i feel sad about the situation too).

**Comment by pedroigor** [comment]
@splatch Your changes are included here https://github.com/keycloak/keycloak/pull/7731.

**Comment by splatch** [comment]
@pedroigor Thank you very much sir! I hope to have chance to test changes somewhere in coming weeks!

**Pull request closed without merging**